### PR TITLE
Cmake refactor deps

### PR DESCRIPTION
**Pull request by petersteneteg**

### CODE REVIEW

File: apps/inviwo/packaging/packaging.cmake [hunk 1]
```diff
@@ -103,15 +101,8 @@ else()
...
-if(NOT APPLE)
-    install(DIRECTORY ${IVW_ROOT_DIR}/data/images           DESTINATION data      COMPONENT images)
-    install(DIRECTORY ${IVW_ROOT_DIR}/data/scripts          DESTINATION data      COMPONENT scripts)
-    install(DIRECTORY ${IVW_ROOT_DIR}/data/volumes          DESTINATION data      COMPONENT volumes)
-    install(DIRECTORY ${IVW_ROOT_DIR}/data/workspaces       DESTINATION data      COMPONENT workspaces)
-
-    install(DIRECTORY ${IVW_ROOT_DIR}/tests/images          DESTINATION tests     COMPONENT images)
-    install(DIRECTORY ${IVW_ROOT_DIR}/tests/volumes         DESTINATION tests     COMPONENT volumes)
-endif()
+install(DIRECTORY ${IVW_ROOT_DIR}/data/  DESTINATION ${IVW_RESOURCE_INSTALL_PREFIX}data  COMPONENT Datasets)
```
Comment (ResearchDaniel) [inline apps/inviwo/packaging/packaging.cmake:104]: This can be other things than datasets such as scripts and transfer functions?

File: CMakeLists.txt [hunk 2]
```diff
@@ -53,18 +47,27 @@ endwhile()
...
 
 #Verify that git submodules has been cloned
 include(cmake/verifysubmodules.cmake)
 verify_submodules("${CMAKE_CURRENT_LIST_DIR}/.gitmodules")
 
 option(BUILD_SHARED_LIBS "Build shared libs, else static libs" ON)
 
-# Needed for find_package(PythonLibsNew 3), which is an improved version of CMake shipped find_pythonLibs
-set(CMAKE_MODULE_PATH "${CMAKE_BINARY_DIR}/pkg/;${CMAKE_MODULE_PATH};${CMAKE_SOURCE_DIR}/ext/pybind11/tools")
+file(REMOVE_RECURSE ${CMAKE_BINARY_DIR}/pkg)
```
Comment (ResearchDaniel) [inline CMakeLists.txt:64]: This seems a bit off, why remove?

File: CMakeLists.txt [hunk 3]
```diff
@@ -53,18 +47,27 @@ endwhile()
...
+set(CMAKE_FIND_PACKAGE_PREFER_CONFIG ON)
+
+list(APPEND CMAKE_MODULE_PATH 
+    ${PROJECT_SOURCE_DIR}/support/cmake/modules
```
Comment (ResearchDaniel) [inline CMakeLists.txt:68]: support/cmake ? Is it moved?

File: CMakeLists.txt [hunk 4]
```diff
@@ -75,30 +78,37 @@ set(CMAKE_MODULE_PATH "${CMAKE_BINARY_DIR}/pkg/;${CMAKE_MODULE_PATH};${CMAKE_SOU
...
-option(IVW_TINY_GLFW_APPLICATION "Build Inviwo Tiny GLFW Application" OFF)
-option(IVW_TINY_QT_APPLICATION   "Build Inviwo Tiny QT Application" OFF)
-
-if(IVW_QT_APPLICATION AND NOT IVW_QT_APPLICATION_BASE)
-    set(IVW_QT_APPLICATION_BASE ON CACHE BOOL "Build base for qt applications. \
-            Used by Qt network editor and Qt minimal application" FORCE)
-    message(STATUS "IVW_QT_APPLICATION_BASE was set to build, due to dependency towards IVW_QT_APPLICATION")
+option(IVW_APP_QTBASE       "Build base for Qt apps. Used by the Qt network editor and the Tiny Qt app" ON)
+option(IVW_APP_INVIWO       "Build Inviwo Qt network editor application" ON)
+option(IVW_APP_MINIMAL_GLFW "Build Inviwo Tiny GLFW Application" OFF)
```
Comment (ResearchDaniel) [inline CMakeLists.txt:92]: Can these be moved to respective app? It would cleanup this file a bit and make it easier for people to add their own

File: CMakeLists.txt [hunk 5]
```diff
@@ -75,30 +78,37 @@ set(CMAKE_MODULE_PATH "${CMAKE_BINARY_DIR}/pkg/;${CMAKE_MODULE_PATH};${CMAKE_SOU
...
+        "Build base for qt applications. Used by Qt network editor and Qt minimal application" 
+        FORCE
+    )
+    message(STATUS "IVW_APP_QTBASE was set to build, due to dependencies")
 endif()
-ivw_enable_modules_if(IVW_QT_APPLICATION QtWidgets)
-ivw_enable_modules_if(IVW_INTEGRATION_TESTS GLFW Base)
-ivw_enable_modules_if(IVW_TINY_GLFW_APPLICATION GLFW)
+
+ivw_enable_modules_if(IVW_APP_INVIWO QtWidgets)
```
Comment (ResearchDaniel) [inline CMakeLists.txt:104]: Same here, can these be moved to respective app?

File: CMakeLists.txt [hunk 6]
```diff
@@ -75,30 +78,37 @@ set(CMAKE_MODULE_PATH "${CMAKE_BINARY_DIR}/pkg/;${CMAKE_MODULE_PATH};${CMAKE_SOU
...
+ivw_enable_modules_if(IVW_APP_MINIMAL_QT QtWidgets)
+ivw_enable_modules_if(IVW_APP_MINIMAL_GLFW GLFW)
+ivw_enable_modules_if(IVW_APP_INVIWO_DOME SGCT)
+ivw_enable_modules_if(IVW_APP_PYTHON Python3 Python3Qt QtWidgets)
+
+option(IVW_TEST_INTEGRATION_TESTS "Build inviwo integration test" ON)
```
Comment (ResearchDaniel) [inline CMakeLists.txt:110]: I think that IVW_TESTS would be better. TEST can be interpreted as unfinished stuff (e.g, we are testing some new functionality).
This would then be IVW_TESTS_INTEGRATION

File: CMakeLists.txt [hunk 7]
```diff
@@ -112,52 +122,19 @@ endif()
...
-
-option (IVW_USE_EXTERNAL_IMG "Link with external image libraries (zlib, png, jpeg, tiff) instead of building them." OFF)
-if (NOT IVW_USE_EXTERNAL_IMG)
-  add_subdirectory(ext/zlib-1.2.11) # Defines target ZLIB::ZLIB
-  add_subdirectory(ext/libpng)      # Defines target PNG::PNG
-  add_subdirectory(ext/libjpeg)     # Defines target JPEG::JPEG
-  add_subdirectory(ext/tiff)        # Defines target TIFF::TIFF
+add_subdirectory(ext)                        # Add external libraries.
+if(IVW_TEST_INTEGRATION_TESTS OR IVW_TEST_UNIT_TESTS)
+    add_subdirectory(tests/testutil)         # Add test utils
```
Comment (ResearchDaniel) [inline CMakeLists.txt:127]: Can this be moved to tests?

File: CMakeLists.txt [hunk 8]
```diff
@@ -112,52 +122,19 @@ endif()
...
+endif()                    
+ivw_register_modules(all_modules)            # Add modules
+add_subdirectory(tools/meta)                 # Add Inviwo meta
+add_subdirectory(src/qt)                     # Add Qt
+add_subdirectory(apps)                       # Add applications
+ivw_add_external_projects()                  # Add external projects
+if(IVW_TEST_INTEGRATION_TESTS)
```
Comment (ResearchDaniel) [inline CMakeLists.txt:134]: How about adding the tests directory and doing this logic there?

File: CMakeLists.txt [hunk 9]
```diff
@@ -112,52 +122,19 @@ endif()
...
-  add_subdirectory(ext/zlib-1.2.11) # Defines target ZLIB::ZLIB
-  add_subdirectory(ext/libpng)      # Defines target PNG::PNG
-  add_subdirectory(ext/libjpeg)     # Defines target JPEG::JPEG
-  add_subdirectory(ext/tiff)        # Defines target TIFF::TIFF
+add_subdirectory(ext)                        # Add external libraries.
+if(IVW_TEST_INTEGRATION_TESTS OR IVW_TEST_UNIT_TESTS)
+    add_subdirectory(tests/testutil)         # Add test utils
+endif()                    
+ivw_register_modules(all_modules)            # Add modules
+add_subdirectory(tools/meta)                 # Add Inviwo meta
```
Comment (ResearchDaniel) [inline CMakeLists.txt:130]: These comments are not very helpful :) 
Not part of this PR but you could consider changing or removing them

File: apps/inviwo/CMakeLists.txt [hunk 10]
```diff
@@ -84,47 +83,23 @@ elseif(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
         MACOSX_BUNDLE_COPYRIGHT "C) 2012-${YEAR} Inviwo Foundation"
     )
     #Following http://www.cmake.org/Wiki/BundleUtilitiesExample
-    install(
```
Comment (ResearchDaniel) [inline apps/inviwo/CMakeLists.txt:87]: Can this be removed?

File: cmake/compileoptions.cmake [hunk 11]
```diff
@@ -62,7 +62,7 @@ function(ivw_define_standard_properties)
             list(APPEND comp_opts "/wd4251") # needs dll-interface   https://msdn.microsoft.com/en-us/library/esew7y1w.aspx
             list(APPEND comp_opts "/wd4505") # unreferenced funtion  https://msdn.microsoft.com/en-us/library/mt694070.aspx
             list(APPEND comp_opts "/w35038") # class member reorder
-            if(NOT IVW_OPENMP_ON)
+            if(NOT IVW_USE_OPENMP)
```
Comment (ResearchDaniel) [inline cmake/compileoptions.cmake:65]: IVW_CFG_USE_OPENMP

File: cmake/installutils.cmake [hunk 12]
```diff
@@ -27,9 +27,14 @@
 # 
 #################################################################################
 
+# Components
+# Application   All the executables and dll etc needed to run the att
```
Comment (ResearchDaniel) [inline cmake/installutils.cmake:31]: att->app

File: cmake/installutils.cmake [hunk 13]
```diff
@@ -27,9 +27,14 @@
...
+# Development   Lib files, pdb files, headers, 
+# Datasets      General datasets i.e. boron etc.
```
Comment (ResearchDaniel) [inline cmake/installutils.cmake:33]: Also other things

**Comment by ResearchDaniel** [comment]
Had some comments that probably apply to follow up PR

**Comment by ResearchDaniel** [comment]
@petersteneteg can you see them?

**Comment by petersteneteg** [comment]
Can not find any :(

**Review by ResearchDaniel — COMMENTED** [review]
Start of review before the PR closed :)

File: apps/inviwo/packaging/packaging.cmake (hunk 1 above)
Comment (petersteneteg) [reply apps/inviwo/packaging/packaging.cmake:104]: Yes, but why are they in the "data" folder then?

File: CMakeLists.txt (hunk 2 above)
Comment (petersteneteg) [reply CMakeLists.txt:64]: I need to remove the old ones, if one is no longer building them

File: CMakeLists.txt (hunk 3 above)
Comment (petersteneteg) [reply CMakeLists.txt:68]: tried to figure out where that one came from, should be removed... good find

File: CMakeLists.txt (hunk 4 above)
Comment (petersteneteg) [reply CMakeLists.txt:92]: no we need them here to enable the modules for them (ivw_enable_modules_if) The alternative could maybe be that we just asset in the app that the needed modules are enabled and then you have to manually add them.
Which is more or less what you would do for you own app at the moment, but it would make the regular use case a bit less easy...

File: CMakeLists.txt (hunk 5 above)
Comment (petersteneteg) [reply CMakeLists.txt:104]: that would not work

File: CMakeLists.txt (hunk 6 above)
Comment (petersteneteg) [reply CMakeLists.txt:110]: Would you do IVW_TESTS_UNIT then? that does not seem right to me

File: CMakeLists.txt (hunk 7 above)
Comment (petersteneteg) [reply CMakeLists.txt:127]: I prefer having the check on the "outside" of the add_subdirectory. Think that makes it easier to understand.

File: cmake/compileoptions.cmake (hunk 11 above)
Comment (petersteneteg) [reply cmake/compileoptions.cmake:65]: We have IVW_USE_* for enabling various dependences, like sigar, openmp, openexr, and IVW_USE_EXTERNAL_* for defining where to get it

File: cmake/installutils.cmake (hunk 13 above)
Comment (petersteneteg) [reply cmake/installutils.cmake:33]: true

File: cmake/installutils.cmake (hunk 12 above)
Comment (petersteneteg) [reply cmake/installutils.cmake:31]: fixed

File: apps/inviwo/CMakeLists.txt (hunk 10 above)
Comment (petersteneteg) [reply apps/inviwo/CMakeLists.txt:87]: ivw_default_install_targets(inviwo) does that for all platforms

File: CMakeLists.txt (hunk 8 above)
Comment (petersteneteg) [reply CMakeLists.txt:134]: the testutils has to be added before the modules and the integration tests after the modules...

File: CMakeLists.txt (hunk 9 above)
Comment (petersteneteg) [reply CMakeLists.txt:130]: true
